### PR TITLE
Add event-flow visualization

### PR DESCRIPTION
**Pull request by williaster**
@mistercrunch @graceguo-supercat 

This PR adds an event flow visualization for analysis of event sequences. The vis is mostly built in [`@data-ui`](https://github.com/williaster/data-ui/tree/chris--event-flow) with [@vx](https://github.com/hshoff/vx) and there is a light wrapper for Superset integration. 

Overall the vis supports the following:
- Takes raw user (or more generically 'entity') event sequences and computes aggregate paths (shown in top panel of vis)
- Allows alignment of event sequences by an arbitrary event **index** and event **type** (e.g., the 2nd click event)
- Allows selection of arbitrary subtrees of the aggregated graph; this filters to the selected subtree in the aggregate panel and shows the corresponding/selected raw event sequences in a bottom panel
- Gives a breakdown of event type counts in the right panel; by clicking on the legend you can filter to / filter out specific types of events.
- X-axis: by default aggregate nodes are positioned according to the mean elapsed time from the previous node, for closely-spaced events you can also position by sequence _number_ (1st, 2nd, 3rd, etc)
- Node sorting: by default nodes are ordered top -> bottom based on high -> low event count meaning most common nodes appear at the top; you can also order by short -> long elapsed time to the next event.

event-flow-demo

note that builds, etc. will fail until the `@data-ui/event-flow` package is published (which I should get to later today), I just wanted to get some 👀 s on the PR

### CODE REVIEW

**Comment by williaster** [comment]
and ere are some pics in Superset:

**Explore controls**


**Dashboard**

**Comment by mistercrunch** [comment]
The image should be square, ideally 1024*1024 to render well

**Comment by mistercrunch** [comment]
This is awesome btw! I'll try to review soon but I'm a little swamped at the moment.

File: superset/assets/visualizations/EventFlow.jsx [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import ReactDOM from 'react-dom';
+
+import {
+  App,
+  withParentSize,
+  cleanEvents,
+  TS,
+  EVENT_NAME,
+  ENTITY_ID,
+} from '@data-ui/event-flow';
```
Comment (mistercrunch) [inline superset/assets/visualizations/EventFlow.jsx]: what does '@%' do? where does the event-flow code live?

Comment (williaster) [inline superset/assets/visualizations/EventFlow.jsx]: it's just a name-spaced package on npm, under the ["@data-ui" user](https://www.npmjs.com/~data-ui) but mainly so all the `@data-ui` packages are in the same repo.

itwill live at [https://github.com/williaster/data-ui/packages/event-flow](https://github.com/williaster/data-ui/tree/chris--event-flow) but I haven't quite merged the chart into the master @data-ui repo, adding one more feature. [this is the (massive 🙈) PR](https://github.com/williaster/data-ui/compare/chris--event-flow?expand=1)

**Comment by williaster** [comment]
The package is published now so the build should be fixed (you can demo it [here](https://williaster.github.io/data-ui/?selectedKind=event-flow&selectedStory=hundredUsers&full=0&down=1&left=1&panelRight=0&downPanel=kadirahq%2Fstorybook-addon-actions%2Factions-panel)).

For visual clarity + performance, with larger number of events expected with use in Superset (10,000s) I also added functionality to "trim" nodes in the visualization based on a minimum event count threshold. This means that sequences are aggregated together but "leaf" nodes are hidden from view.

**Before / after trimming**
event-flow-event-filter

File: superset/assets/javascripts/explore/stores/controls.jsx [hunk 2]
```diff
@@ -1273,5 +1273,11 @@ export const controls = {
     hidden: true,
     description: 'The number of seconds before expiring the cache',
   },
+
+  generic_checkbox: {
```
Comment (mistercrunch) [inline superset/assets/javascripts/explore/stores/controls.jsx]: This is the name that will be use in the saved slice's json. Having something generic here will make it harder if/when trying to generate slice dynamically, and harder to read the json and understand what that maps to. Also references in the backend/frontend code won't be as explicit. Let's use a name that describe the actual feature this triggers.

File: superset/assets/package.json [hunk 3]
```diff
@@ -56,7 +57,7 @@
     "jquery": "^3.2.1",
     "jsdom": "9.12.0",
     "lodash.throttle": "^4.1.1",
-    "mapbox-gl": "^0.26.0",
+    "mapbox-gl": "^0.27.0",
```
Comment (mistercrunch) [inline superset/assets/package.json]: I'm guessing this was an attempt at fixing the `npm install` bug. It's been fixed on master by using `yarn` instead of `npm install` so this is probably not needed anymore. Though if you've tested that the mapbox viz renders alright I'm not against the bump here even though it would belong in another PR ideally.

File: superset/assets/visualizations/EventFlow.jsx [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+/*
+ * This function takes the slice object and json payload as input and renders a
+ * responsive <EventFlow /> component using the json data.
+ */
+function renderEventFlow(slice, json) {
+  const container = document.querySelector(slice.selector);
+  const hasData = json.data && json.data.length > 0;
+
+  // the slice container overflows ~80px in explorer, so we have to correct for this
+  const isExplorer = (/explore/).test(window.location.pathname);
```
Comment (mistercrunch) [inline superset/assets/visualizations/EventFlow.jsx]: not sure if I understand why, but I'm wondering if using `slice.width()` and `slice.height()` would address the 80px issue?

**Review by mistercrunch — CHANGES_REQUESTED** [review]
Sorry for the delay on the review, but this is really amazing and the kind of visualization that is a key "selling" point for Superset. I have a few notes for otherwise LGTM!

I'm curious about your CSS in your module and want to confirm that it's namespace alright so it doesn't spill on other visualization (the same way treemap did on yours).

File: superset/assets/visualizations/EventFlow.jsx (hunk 4 above)
Comment (williaster) [inline superset/assets/visualizations/EventFlow.jsx]: I tried this to begin with but it doesn't seem to work/couldn't come up with any fix for the css. The panel dom looks like:
```html
<div class="panel">
  <div class="panel-heading">...</div>
  <div class="panel-body">...</div>
</div>
```
`panel-body` is what the slice is in and it seems to have the same height as the parent `panel`. that means it overflows by the height of `panel-heading` :/

**Comment by williaster** [comment]
@mistercrunch I made the following changes

- removed mapbox version bumps (these broke the build for a while, looks good now 👌)
- updated the thumbnail png to 1024 x 1024
- made the checkbox more specific to event-flow

**Comment by mistercrunch** [comment]
I fixed the coveralls token in master, please rebase+push

**Comment by williaster** [comment]
@mistercrunch done

**Comment by williaster** [comment]
let me know if I should add a test for the coverage.

**Comment by vnnw** [comment]
Can you add a demo into Superset distribution?

**Comment by williaster** [comment]
@vnnw yes this would be a good addition to get people up and going a little more easily.
